R/DBConnection.R

```diff
@@ -106,6 +106,97 @@ setGeneric("dbSendQuery",
   valueClass = "DBIResult"
 )
 
+#' Sentinel value to indicate the early termination of
+#' \code{\link{dbGetChunkedQuery}}
+#' @export
+dbBreak <- structure(list(), class = "dbi_break_sentinel")
+
+dbFetchChunked <- function(rs, n, callback) {
+  while (TRUE) {
+    chunk <- dbFetch(rs, n = n)
+    if (nrow(chunk) == 0) {
+      return(chunk)
+    }
+    result <- callback(chunk)
```

Comment:
I think Hadley meant `callback(chunk, index = rowsSoFar)` so that the callback can know how far it is into the iteration.